src/tools/non-promotion-reasons.md

```diff
@@ -1,45 +1,380 @@
+  while (true) {    // (1)
+    print(p.value); // (2) ERROR
+    var next = p.next;
+    if (next == null) break;
+    p = next;       // (3)
+  }
+}
+{% endprettify %}
+
+When flow analysis reaches (1),
+it looks ahead and sees the write to `p` at (3).
+But because it's looking ahead,
+it hasn't yet figured out the type of the right-hand side of the assignment,
+so it doesn't know whether it's safe to retain the promotion.
+To be safe, it invalidates the promotion.
+
+You might fix this problem by moving the null check to the top of the loop:
+
+{:.good}
+{% prettify dart tag=pre+code %}
+f(Link? p) {
+  while (p != null) {
+    print(p.value);
+    p = p.next;
+  }
+}
+{% endprettify %}
+
+This situation can also arise in `switch` statements if
+there's a `case` block that has a label,
+because you can use labeled `switch` statements to construct loops:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+f(int i, int? j, int? k) {
+  if (j == null) return;
+  switch (i) {
+    label:
+    case 0:
+      print(j.isEven);
+      j = k;
+      continue label;
+  }
+}
+{% endprettify %}
+
+
+### In catch after possible write in try
+
+**The cause:**
+The variable might have been written to in a `try` block,
+and execution is now in a `catch` block.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+f(int? i, int? j) {
+  if (i == null) return;
+  try {
+    i = j;                 // (1)
+    if (i == null) return; // (2)
+  } catch (...) {
+    print(i.isEven);       // ERROR
+  }
+}
+{% endprettify %}
+
+In this case, flow analysis doesn't consider `i.isEven` safe,
+because it has no way of knowing when in the `try` block
+the exception might have occurred,
+so it conservatively assumes that it might have happened between (1) and (2),
+when `i` was potentially `null`.
+
+Similar situations can occur between `try` and `finally` blocks, and
+between `catch` and `finally` blocks.
+
+Because of a historical artifact of how the implementation was done,
+these try/catch/finally situations don't take into account
+the right-hand side of the assignment,
+similar to what happens in loops.
+
+***[PENDING: show fix?]***
+
+
+### Subtype mismatch
+
+**The cause:**
+The type you're trying to promote to isn't a subtype of
+the variable's current promoted type
+(or wasn't a subtype at the time of the promotion attempt).
+
+Example:
+
+{:.good}
+{% prettify dart tag=pre+code %}
+f(Object o) {
+  if (o is Comparable) {             // (1)
+    if (o is Pattern) {              // (2)
+      print(o.matchAsPrefix('foo')); // (3) ERROR
+    }
+  }
+}
+{% endprettify %}
+
+In this example, `o` is promoted to `Comparable` at (1), but
+it's not promoted to `Pattern` at (2),
+because `Pattern` is not a subtype of `Comparable`.
+(The rationale is that if we did promote,
+then you wouldn't be able to access methods on `Comparable` anymore).
+Note that just because `Pattern` isn't a subtype of `Comparable`
+doesn't mean the code at (3) is dead;
+if you define a class that
```

Comment:
Aha!  I just realized that the standard library already has a class that implements both `Comparable` and `Pattern`: the `String` class!  So if you want you could change this sentence to something like:

> Note that just because `Pattern` isn't a subtype of `Comparable` doesn't mean the code at (3) is dead; `o` could be a `String` (which implements both `Comparable` and `Pattern`).